internal/mesh/internal/controllers/routes/mapper.go

```diff
@@ -0,0 +1,213 @@
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v2beta1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v2beta1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func MapHTTPRoute(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	if !types.IsHTTPRouteType(res.Id.Type) {
+		return nil, fmt.Errorf("type is not a http route type: %s", res.Id.Type)
+	}
+	return dependency.MapDecoded(mapXRoute[*pbmesh.HTTPRoute])(ctx, rt, res)
+}
+
+func MapGRPCRoute(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	if !types.IsGRPCRouteType(res.Id.Type) {
+		return nil, fmt.Errorf("type is not a grpc route type: %s", res.Id.Type)
+	}
+	return dependency.MapDecoded(mapXRoute[*pbmesh.GRPCRoute])(ctx, rt, res)
+}
+
+func MapTCPRoute(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	if !types.IsTCPRouteType(res.Id.Type) {
+		return nil, fmt.Errorf("type is not a tcp route type: %s", res.Id.Type)
+	}
+	return dependency.MapDecoded(mapXRoute[*pbmesh.TCPRoute])(ctx, rt, res)
+}
+
+func mapXRoute[T types.XRouteData](_ context.Context, _ controller.Runtime, xr *resource.DecodedResource[T]) ([]controller.Request, error) {
+	refs := parentRefSliceToRefSlice(xr.Data.GetParentRefs())
+	return controller.MakeRequests(pbmesh.ComputedRoutesType, refs), nil
+}
+
+func MapServiceNameAligned(
+	ctx context.Context,
+	rt controller.Runtime,
+	res *pbresource.Resource,
+) ([]controller.Request, error) {
+	// Since this is name-aligned, just switch the type and find routes that
+	// will route any traffic to this destination service.
+	svcID := resource.ReplaceType(pbcatalog.ServiceType, res.Id)
+
+	svc, err := rt.Cache.Get(pbcatalog.ServiceType, "id", svcID)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: this may over-trigger slightly and include extraneous
+	// re-reconciles due to also traversing the failover policies backwards in
+	// addition to the stuff we want.
+	//
+	// This is not a correctness-problem.
+	return MapService(ctx, rt, svc)
+}
+
+func MapService(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	if !types.IsServiceType(res.Id.Type) {
+		return nil, fmt.Errorf("type is not a service type: %s", res.Id.Type)
+	}
+
+	return dependency.MultiMapper(
+		dependency.ReplaceType(pbmesh.ComputedRoutesType), // TODO: may be unnecessary
+		dependency.MapperWithTransform(
+			// (2) find xRoutes with the provided service as a backend; and enumerate all parent refs
+			mapBackendServiceToComputedRoutes,
+			// (1) find failover policies that include this as a destination; also include itself
+			transformServiceToBackendServices,
+		),
+	)(ctx, rt, res)
+}
```

Comment:
This feels like it could be replaced with more static mapper + transform:

```
mapService := dependency.MultiMapper(
   dependency.ReplaceType(pbmesh.ComputedRoutesType), // TODO: may be unnecessary
   dependency.MapperWithTransform(
 	   // (2) find xRoutes with the provided service as a backend; and enumerate all parent refs
      mapBackendServiceToComputedRoutes,
      // (1) find failover policies that include this as a destination; also include itself
      transformServiceToBackendServices,
   ),
)

mapServiceNameAligned := dependency.MapperWithTransform(
   mapService,
   transformNameAlignedService,
)

func transformNameAlignedService(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]*pbresource.Resource, error) {
   // Since this is name-aligned, just switch the type and find routes that
 	// will route any traffic to this destination service.
 	svcID := resource.ReplaceType(pbcatalog.ServiceType, res.Id)

 	svc, err := rt.Cache.Get(pbcatalog.ServiceType, "id", svcID)
 	if err != nil {
 		return nil, err
 	}

 	// NOTE: this may over-trigger slightly and include extraneous
 	// re-reconciles due to also traversing the failover policies backwards in
 	// addition to the stuff we want.
 	//
 	// This is not a correctness-problem.
   return []*pbresource.Resource{svc}, nil
}
```

The advantage would be not creating new functions and closures all the time just to invoke them a single time.

Side Note: I just now realized that `MapperWithTransform` would likely be better if the args were swapped (transform first and then mapper) so that in the code it seems like a more natural progression of the pipeline)